CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/daemon/QualityControlAutoVetter.java

```diff
@@ -142,71 +148,20 @@ public void graphChanged(final GraphChangeEvent event) {
-            final Thread thread = new Thread("Quality Control View: Run Rule") {
-                @Override
-                public void run() {
-                    final int selectedAttribute = VisualConcept.VertexAttribute.SELECTED.get(graph);
-                    final int identifierAttribute = VisualConcept.VertexAttribute.IDENTIFIER.get(graph);
-                    final int typeAttribute = AnalyticConcept.VertexAttribute.TYPE.get(graph);
-                    if (selectedAttribute != Graph.NOT_FOUND && identifierAttribute != Graph.NOT_FOUND && typeAttribute != Graph.NOT_FOUND) {
-                        final int vxCount = graph.getVertexCount();
-                        for (int position = 0; position < vxCount; position++) {
-                            final int vertex = graph.getVertex(position);
-                            final String identifier = graph.getStringValue(identifierAttribute, vertex);
-                            final SchemaVertexType type = graph.getObjectValue(typeAttribute, vertex);
-
-                            final boolean selected = graph.getBooleanValue(selectedAttribute, vertex);
-                            if (selected) {
-                                vertexList.add(vertex);
-                                identifierList.add(identifier);
-                                typeList.add(type);
-                            }
-                        }
-                    }
-
-                    // Set up and run each rule.
-                    if (!vertexList.isEmpty()) {
-                        for (final QualityControlRule rule : getRules()) {
-                            rule.clearResults();
-                            rule.executeRule(graph, vertexList);
-                            registeredRules.add(rule);
-                        }
-                    }
-
-                    countDownLatch.countDown();
-                }
-            };
-            thread.start();
-
-            try {
-                countDownLatch.await();
-            } catch (InterruptedException ex) {
-                Exceptions.printStackTrace(ex);
-                Thread.currentThread().interrupt();
-            }
-        }
-
-        // Build quality control events based on results of rules.
-        // Sort by descending risk.
-        final List<QualityControlEvent> qualityControlEvents = new ArrayList<>();
-        for (int i = 0; i < vertexList.size(); i++) {
-            final QualityControlEvent qualityControlEvent = new QualityControlEvent(vertexList.get(i), identifierList.get(i), typeList.get(i), Collections.unmodifiableList(registeredRules));
-            qualityControlEvents.add(qualityControlEvent);
+    public static void updateQualityControlState(final GraphReadMethods graph) {
+        final Graph currentGraph = GraphManager.getDefault().getActiveGraph();
+        Future<?> stateFuture = PluginExecution.withPlugin(new QualityControlViewStateUpdater(graph))
+                .executeLater(currentGraph);
```

Comment:
For some background, there is a quality score available from the data access view near the Go button. An analyst is meant to do their own checks and also double check the quality score before hitting go. The only risk here is that given the plugin runs "later", there is no guarantee that it will finish before the analyst decides to press Go.

The only advantage of this stuff running on the EDT (though obviously very bad design) is that there is less chance for this race condition to occur because Constellation would effectively lock up until the quality score was updated.

I will test this out in anger and if you haven't done so already, I will see if there is a way to show a progress or "thinking" graphic to the little quality score section to let the analyst know that its not ready for you to press Go.

Hmm, thinking out aloud, perhaps the Go button could also be disabled whilst these jobs run?

Reply:
I've been thinking of how to disable the Go button depending on if the scores are updating. The only solution I could come up with would be to have a static boolean variable within DAV pane. So when QCV updates the rules, the DAV would know when to change button state. This leads to one issue, I don't know how I would be able to trigger a disable or enable of the button from the static context. 
I'm all ears for ideas, as I don't think my solution will exactly work in practice.

Reply:
@aldebaran30701 How about having a static method in the DAV that you can call to enable or disable it from the QC code?

Reply:
That's what I was trying to explain, but I wouldn't think I would have access to the instance of the button to change it's state (enabled or disabled) because of the static context of that method.

Reply:
Can you try making the button static?

Reply:
I've been messing around with it but cannot get around a cyclic dependency when trying to use the static method ```DataAccessPane.setCanRun()``` - A static method which I made.
When I try to access ```DataAccessPane``` from QCV AutoVetter class it makes the cyclic dependency.

Reply:
Thanks for trying this out. Yeah circular dependencies is a problem and not too surprised it came up because of how the widget is embedded in the DAV.

One way to get around this would be to introduce a mini event driven design pattern so that you can "fire" and event that can tell the button to be enabled or disabled. You can fire the event from the QC code and pick it up from the DAV code without the dependency.

There are several places where this is being done and for an example see `fireChangeEvent()` in `PluginParameter.` Happy for you to have a go with this if you like and if you need any assistance let me know.

Reply:
From memory, the DAV depends on the QCV (as it makes use of that widget) - so you can't also have the QCV depend on the DAV as that would introduce a circular dependency. What you want instead is to treat the QCV widget as a service - so it would update itself and provide information about its state for the DAV (or any other clients) to use as necessary.

Reply:
So from your comment @cygnus-x-1 would that mean I would change code in QCV to become a service, use the fire change event pattern @arcturus2 described, or are they the same thing?

Reply:
I think they're essentially the same - you'd be sending a signal from the QCV that something has happened, and it would be up to the DAV to listen for that and respond.

Reply:
I've implemented an observer to fire events and let the button update dependent on the rules running. Let me know if you notice anything. Should I be calling ```removeListener()``` at any point?

Reply:
Yes @aldebaran30701 for sure and good on you for thinking thoroughly about this part. The count of addListener's should equal the removeListener's or else that is by definition a memory leak.

I would suggest the removeListener should run when the component is closed. Therefore can you consider moving the `addListener()` to `DataAccessViewTopCoponent().handleComponentOpened()` and the remove in the `handleComponentClosed()`. I think you have access to the `DataAccessPane` from there so it should be OK to do. Would you agree?

Reply:
I've moved the listeners to the specified locations, and fixed the concurrent modification exception.

Reply:
Implemented in most recent commit